titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java

```diff
@@ -95,7 +95,10 @@
                     if (responseHeaders.header(dockerDigestHeaderKey).isEmpty()) {
                         return Mono.error(new TitusRegistryException(TitusRegistryException.ErrorCode.MISSING_HEADER, "Missing required header " + dockerDigestHeaderKey));
                     }
-                    return Mono.just(responseHeaders.header(dockerDigestHeaderKey).get(0));
+                    return response.toEntity(String.class).flatMap(responseEntity -> {
```

Comment:
docs from the `CilentResponse` interface:

```
 * <p><strong>NOTE:</strong> When given access to a {@link ClientResponse},
 * through the {@code WebClient}
 * {@link WebClient.RequestHeadersSpec#exchange() exchange()} method,
 * you must always use one of the body or toEntity methods to ensure resources
 * are released and avoid potential issues with HTTP connection pooling.
 * You can use {@code bodyToMono(Void.class)} if no response content is
 * expected. However keep in mind that if the response does have content, the
 * connection will be closed and will not be placed back in the pool.
```

That means any of the cases above this could also leak a connection. I suggest we rewrite this as:

```diff
diff --git a/titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java b/titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java
index 704f19df..96f44d97 100644
--- a/titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java
+++ b/titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java
@@ -20,6 +20,7 @@ import java.time.Duration;
 import java.util.AbstractMap;
 import java.util.Collections;
 import java.util.Map;
+import java.util.Objects;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.inject.Inject;
@@ -32,6 +33,7 @@ import com.netflix.titus.common.util.guice.annotation.ProxyConfiguration;
 import io.netty.handler.codec.http.HttpResponseStatus;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpHeaders;
 import org.springframework.web.reactive.function.client.ClientResponse;
 import org.springframework.web.reactive.function.client.WebClient;
 import reactor.core.publisher.Mono;
@@ -78,24 +80,25 @@ public class DefaultDockerRegistryClient implements RegistryClient {
         return restClient.get().uri(buildRegistryUri(repository, reference))
                 .headers(consumer -> headers.forEach(consumer::add))
                 .exchange()
+                .flatMap(response -> response.toEntity(String.class))
                 .flatMap(response -> {
-                    if (response.statusCode().value() == HttpResponseStatus.NOT_FOUND.code()) {
+                    if (response.getStatusCodeValue() == HttpResponseStatus.NOT_FOUND.code()) {
                         return Mono.error(
                                 new TitusRegistryException(TitusRegistryException.ErrorCode.IMAGE_NOT_FOUND,
                                         String.format("Image %s:%s does not exist in registry", repository, reference))
                         );
                     }
-                    if (!response.statusCode().is2xxSuccessful()) {
+                    if (!response.getStatusCode().is2xxSuccessful()) {
                         return Mono.error(
                                 new TitusRegistryException(TitusRegistryException.ErrorCode.INTERNAL,
-                                        String.format("Cannot fetch image %s:%s metadata: statusCode=%s", repository, reference, response.statusCode()))
+                                        String.format("Cannot fetch image %s:%s metadata: statusCode=%s", repository, reference, response.getStatusCode()))
                         );
                     }
-                    ClientResponse.Headers responseHeaders = response.headers();
-                    if (responseHeaders.header(dockerDigestHeaderKey).isEmpty()) {
+                    HttpHeaders responseHeaders = response.getHeaders();
+                    if (!responseHeaders.containsKey(dockerDigestHeaderKey)) {
                         return Mono.error(new TitusRegistryException(TitusRegistryException.ErrorCode.MISSING_HEADER, "Missing required header " + dockerDigestHeaderKey));
                     }
-                    return Mono.just(responseHeaders.header(dockerDigestHeaderKey).get(0));
+                    return Mono.just(Objects.requireNonNull(responseHeaders.getFirst(dockerDigestHeaderKey)));
                 })
                 .timeout(Duration.ofMillis(titusRegistryClientConfiguration.getRegistryTimeoutMs()))
                 .retryWhen(TitusWebClientAddOns.retryer(
```

That way we ensure the body is always drained and connections are returned to the pool.

Reply:
Also, docs on the `.exchange()` method:

```
		 * Perform the HTTP request and return a {@link ClientResponse} with the
		 * response status and headers. You can then use methods of the response
		 * to consume the body:
		 * <p><pre>
		 * Mono&lt;Person&gt; mono = client.get()
		 *     .uri("/persons/1")
		 *     .accept(MediaType.APPLICATION_JSON)
		 *     .exchange()
		 *     .flatMap(response -&gt; response.bodyToMono(Person.class));
		 *
		 * Flux&lt;Person&gt; flux = client.get()
		 *     .uri("/persons")
		 *     .accept(MediaType.APPLICATION_STREAM_JSON)
		 *     .exchange()
		 *     .flatMapMany(response -&gt; response.bodyToFlux(Person.class));
		 * </pre>
		 * <p><strong>NOTE:</strong> You must always use one of the body or
		 * entity methods of the response to ensure resources are released.
		 * See {@link ClientResponse} for more details.
```

Reply:
Thanks. I have made suggested changes.